api/envoy/service/rate_limit_quota/v3/rlqs.proto

```diff
@@ -0,0 +1,130 @@
+syntax = "proto3";
+
+package envoy.service.rate_limit_quota.v3;
+
+import "envoy/type/v3/token_bucket.proto";
+
+import "google/protobuf/duration.proto";
+import "google/protobuf/timestamp.proto";
+import "google/protobuf/wrappers.proto";
+
+import "xds/annotations/v3/status.proto";
+
+import "udpa/annotations/status.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.service.rate_limit_quota.v3";
+option java_outer_classname = "RlqsProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/service/rate_limit_quota/v3;rate_limit_quotav3";
+option java_generic_services = true;
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+option (xds.annotations.v3.file_status).work_in_progress = true;
+
+// [#not-implemented-hide:]
+// [#protodoc-title: Rate Limit Quota Service (RLQS)]
+
+service RateLimitQuotaService {
+  rpc StreamRateLimitQuotas(stream RateLimitQuotaUsageReports)
+      returns (stream RateLimitQuotaAssignments) {
+  }
+}
+
+message RateLimitQuotaUsageReports {
+  // [#next-free-field: 7]
+  message BucketQuotaUsage {
+    // Bucket id for which request quota usage is reported.
+    BucketId bucket_id = 1 [(validate.rules).message = {required: true}];
+
+    // Auto-incrementing number used to deduplicate usage reports.
+    uint64 report_id = 2;
```

Comment:
I had three cases in mind:
1) Idempotence. F.e. there's a disconnect in the middle of the RPC, and the client is able to retry the message. The client doesn't need know if the server processed the initial message, because the server tracks the last accepted report_id, and can skip/ack the report.
2) Debugging. It'll be easier identify the case when due to faulty client code messages are send out-of-order.
3) I was considering to use this field set to `null` (using the wrapped type) to indicate the client is unsubscribing from a particular bucket.

Reply:
1. Having a report ID doesn't seem useful for idempotence, because the server will likely not be a single process anyway, and if the client reconnects, it may not hit the same instance it was previously connected to.  So the only way for the server to use this is to build a bunch of infrastructure to store the last-seen report ID for every single individual client talking to any server instance.  That seems like overkill.  In general, I think we just accept that there will be a bit of reporting churn when the connection is closed and needs to be reopened.
2. I don't see why a client would ever send messages out of order.  In general, it will just have a queue and will send them in order.  This doesn't seem like a likely source of bugs.
3. As we discussed, I think the server should decide to abandon a bucket instead of having the client do it, so this isn't an issue.

Given that reports are going to be sent from all clients for each bucket on a fairly short interval, I think it makes sense to avoid unnecessary overhead here.  I don't see a use-case for this field that is worth the overhead of having it.  I think we should just remove it.

Reply:
Thanks for the clarification Mark. Now I see that the downsides outweigh the benefits. Will get rid of it.